### DISCUSSION TITLE
Add URL Rewrite to Core: Friendly URLs?

### DISCUSSION THREAD

**Discussion by torvista** [zencart/zencart#5992] [Ideas]
Yawn, boring but necessary discussion. Times have changed since the RodG (RIP) discussions.

Native URL:  
https://www.shop.com/index.php?main_page=product_info&cPath=1_5&products_id=8113
URL rewriting:
https://www.shop.com/Widgets/Small Widgets/widget1

Whether you want/use it or not, is it not expected in 2023 that an ecommerce solution offers this functionality?

We have two plugins.
Ultimate
https://github.com/lat9/usu

Ceon
free a: https://github.com/JSWebSteve/Ceon-URI-Mapping-V5.1.0
+ optional add-on (commercial) : https://ceon.net/seo-modules/ceon-uri-mapping-manager

So maintenance time and efforts are going to two places.

Should this functionality be in core?
If so, which should be used as a base? - Opinions from devs with experience of both?

**Comment by dbltoe** [comment]
Both mods are exceptional products by exceptional supporters of Zen Cart.

Both mods are impacted in so many ways when ZC is customized that taking either to the core could involve more time maintaining and be a poor ROI.

Even Moz is still "meh" on the subject.  https://moz.com/learn/seo/url#:~:text=Rankings&text=While%20using%20a%20URL%20that,include%20a%20keyword%20in%20them.

What about the vitamin dealer who has 50 SKUs for Calcium?  How much extra effort is it going to take for the larger store owner to manage a "friendly" URL and keep it where they don't have to consult a spreadsheet each time an item is ordered in order to make sure they send the right product?

You'd think if it were so beneficial, that small company with the swooshing arrow on all their packages would use them.  They do use them in a way

> https://www.BIG_GUY.com/Nature-Made-Calcium-Magnesium-Vitamin/dp/B000OD1F7U?pd_rd_w=0x1eX&content-id=amzn1.sym.1a072a44-84d1-4c51-bf0f-3547594e8217&pf_rd_p=1a072a44-84d1-4c51-bf0f-3547594e8217&pf_rd_r=DSQWZZE97BXP3PBZ3KDZ&pd_rd_wg=MoG2O&pd_rd_r=aca81c85-68a6-48e5-814f-a98d4fe81dc6&pd_rd_i=B000OD1F7U&psc=1&ref_=pd_bap_d_grid_rp_0_1_ec_i

But, you'll note they add a lot of data after the description of the item.  Not so easy to do with ZC if you don't have the IT staff those guys have.  And, that URL is not really friendly.  They just added the folder for the product in the URL.

I know it's somewhat like the bible.  One group will find all the passages in the bible that support drinking alcohol and the other will find just as many references to the evil of alcohol.

We host two sites that beat most of their competition quite handily and one that beats the big guy with the new fiance.  Neither uses other than standard ZC URLs for clarity.

I'm just not sure there's enough of an ROI for the time spent maintaining all facets of the process.  I would certainly put something like Advanced Shipper or a QuickBooks interface above this in priority.  For real owner assistance, Clone a Template is far more beneficial to us in testing and maintenance.

So, I guess I'm more inclined to be in the group that says you might be able to turn water into wine but you still don't have to drink it to have a good time.  😄

**Reply by Motovista** [reply]
You say- "We host two sites that beat most of their competition quite handily and one that beats the big guy with the new fiance. Neither uses other than standard ZC URLs for clarity." How many of these sites are selling something regarding ZenCart?

**Comment by torvista** [comment]
A valuable response, as always.

> Both mods are impacted in so many ways when ZC is customized that taking either to the core could involve more time maintaining and be a poor ROI.

I can only speak as a CEON user, and I haven't seen that at all. Through the years I've updated it for php and to make use of new observers: in the end despite the masses of code/multiple files it only modifies the link generation. I've rarely had to fix it for core updates.

But, as always, I'm banging on about the big picture for Zen Cart...

Irrespective of whether this functionality is of value or not in the real world....
- is it something that ALL the other carts offer out of the box and promote?
- is it therefore something that new users (novices) look for/**think is important** and so reject ZC because it is not included?

Should it be there just because it is expected in 2023?

This is a commercial decision, not a coding one, uncomfortable as it is.

**Reply by Motovista** [reply]
Yes.
Yes.
 and it definitely should be there.

**Comment by webconcept-one** [comment]
I have used CEON for 5 years. I can say that it works well but there was no noticable SEO benefit at all. I'm regarding Search Console data regularly.
I even wrote a post about this topic which got some remarkable traffic.
You may believe when I say that I had been so PRO to Friendly URL years ago.

If you go back from SEO URL to standard there will be a negative impact of course because of lack of rewrite to non-Friendly-URL - of course...

In CEON free package you need to re-save every of your 20.000 (or even many more) of items to create the URL. You can use the "Pro" Version of course. I did that for >2.000 products - and really - there was no measurable impact just because of the Friendly URL.

Friendly URL don't make any difference to SEO, even when the structure is very deep with lots of sub-categories (this has often been a argument PRO Friendly URL). The Crawler doesn't have any problems to index deeper structures.
And if you have lots of sub-categories. Really who wants to navigate through an insane long URL? Breadcrumbs navigation does exactly  this.

Keeping Zen-cart secure, fast, flexible and maintainable should be high priority.
Nice looking features IMHO shouldn't

**Comment by torvista** [comment]
> Friendly URL don't make any difference to SEO

I'm sure you are right. It is purely aesthetic.

My comments about what to add to Zen Cart are usually based on my big-picture viewpoint that the most pressing issue for the project is reversing the declining market-share to get new blood for development.
Part of this must be by giving potential new (un-informed) users what they think they want/need, such as this aspect.
If every other competitor has this, ZC will be classed as missing basic functionality, incorrect as that conclusion is.
I see this functionality as a necessary evil and yes it would be irritating to be forced to include it.
But if a current, maintained plugin works already, why not drop it in and leave it switched off!

Maybe I am wrong, but usually there is no further discussion from anyone who has control input to the project.

**Reply by webconcept-one** [reply]
"Maybe I am wrong, but usually there is no further discussion from anyone who has control input to the project."

I don't understand that. Even translator can't translate that properly. Can you explain?

**Reply by lat9** [reply]
I _believe_ that @torvista is referring to me and the other "members" of this repository who have the authority to accept or decline a PR.

The underlying issue with pulling a plugin into the Zen Cart core is that, once in core, the maintenance responsibility then falls on the the members instead of the original plugin owner.

Since I am the current owner of [Ultimate SEO URLs](https://github.com/lat9/usu) using that as the base would be an easier 'pull into core' than the Ceon URI Mappings.  I'll note also that while USU is currently a drop-in plugin, the Ceon offering has many tentacles into the current core-code.

**Reply by torvista** [reply]
Correctly surmised.

Let's be honest, any plugin maintained by lat9 is good enough code-wise to go into core already and if the subsequent responsibility is shared that's a good thing is it not?

I've been using Ceon for years and was keeping it on life-support for a while, so should have more attachment to it than most, but still, I don't care if something else gets used/it gets retired, such is progress.

**Reply by webconcept-one** [reply]
When you try to exclude anyone from discussion then write it properly please, so your intention can be understood:

"Maybe I am wrong, but usually there is no further discussion from anyone who has **[NO]** control input to the project.

**Reply by torvista** [reply]
Sorry if my meaning was not clear.
And I don't/can't exclude anyone from discussion.
And, my original phrase was correct, as lat9 explained: the progress of this project is determined solely by the very few people who have control (= PR merge permissions), and in the multiple discussions that I have instigated over the years that relate to the project plan/roadmap, those people, with the notable exception of lat9, hardly enter into the discussion when in fact their input has the most weight/no-one else's opinion actually matters.

**Reply by webconcept-one** [reply]
When I'm wrong then because of language barrier.
I want to help to make the best out of Zen-Cart, too

**Reply by torvista** [reply]
No problem, nothing that cannot be resolved with a bit of understanding and courtesy. Something that is occasionally lacking here.
What's important is the participation, which should always be welcomed.

**Comment by harmonyball** [comment]
My first ever comment on GitHub. I'm not a coder - have been with ZC for 12 years + with only 3 active websites now but had more.  My opinion for what it is worth is to forget the SEO aspect, that arguement will always be a draw - the tag 'SEO' added to the plugins and plastered all over the internet is nothing more than that, just a tag, and SEO gets most webmasters attention (User Friendly is a better tag). 

User Friendly URL's are what Customers expect and prefer (as does Google) in 2024 - it all comes down to the Customer and their expectations and their level of comfort when on the site or looking at a link to a site, (they are more comfortable reading friendly url's rather than a bunch of goglydook code). Google calls it Customer Experience.

So can a version of friendly urls be included in Core as radio button selection, on/off (just like it is on many plugins)?

cheers,
Mike

p.s.  I have to agree with Torvista - as a non coder if I was doing now in 2024 what I did in 2011 i.e. investigating which platform may be best for me to commence building my online eCommerce presence then I would not be inclined to give ZenCart a tick on that point, (if I did my homework properly on 'User Friendly URL's' aspect in comparisons).

**Reply by dbltoe** [reply]
Welcome to GitHub and this discussion, Mike!  The more the merrier as we all want ZC to prosper.

I won't dwell again on my earlier reasoning but did want to "freshen" the research I've done.

IMHO, friendly URLs have gone the way of keywords.  Actually, keywords were way more effective until their misuse caused Google and other SEs to drop them from consideration when ranking.

When someone uses Zen Cart for their site, Google and other SEs are all over it immediately.  And, once the site is indexed by the SEs, friendly URLs drop out of the picture for rank.  All they are looking for in a URL is literally the address to go look at your product.  Not unlike the newly married couple looking for a place to live, the address may be simply 2120 First Street or 2120 First Street, Unit 4, Apt 17.  Once they get to the address, they just want to see if the accommodations are as advertised.  Imagine that apartment 17 is advertised as a large one bedroom and the bedroom would not fit a king-sized bed.  They will be telling everyone about the bad description regardless of the unit's location.

So, what the SEs look for after indexing (in order of their importance) is Accessibility, HTML/CSS Validation, and proper links (internal and external).  Next comes page speed and mobile/responsive friendliness.  Even domain age comes into play before friendly URLs once the site is indexed.

A while back, I did [a little bit of research](https://jeandret.com/zenCartStats.pdf) that you may want to take a look at.  This research was done using a load of 1.5.8a with demo items and no other mods.  Home page to home page.  I'm sure some of the other folks have improved in an area or two since my research but, ZC 2.0.0 is looking like it will wind up with higher scores.  This is what needs to be "sold" to the would-be store owner.  What's not on the chart is page load speed.  Some of my testing with 2.0.0 and bootstrap show a desktop speed of 0.5 seconds and a mobile of 2.2 seconds.  The mobile with demo items is slow due to image sizes.

One of our hosting clients has a special shipping situation that forced them to go to Shopify.  They report that their sales are up slightly (ease of shipping being the reason) but the fees ($$ for this and $$ for that) are eating up what they are gaining.

To my mind, shipping and payment methods should be of higher priority.  One customer sells flags and flag poles.  Without Advanced Shipper, they would not be able to manage what goes with which carrier.  Multi-Ship would be of benefit to them if a large automotive dealership chain wanted to send new flags to each of their dealers.  But, again, these are things that will require someone to manage.  While I will always do what I can for Zen Cart, like you I am not well-versed in PHP.

Edit--  No sooner than I hit enter, the following link showed up in my email.  Very interesting.

https://www.semrush.com/blog/ranking-factors-semrush-study/?utm_source=marketo&utm_medium=email&utm_campaign=engage_brand_EduNewsletterAcademy_mass_en_190124_free_newsletter&utm_content=blog&mkt_tok=NTE5LUlJWS04NjkAAAGQwHrR_BwDnRT_HjI6-hBo6_0ND3aHa0IGq6O-HDy0LMyPoKqFWvpOTqdiB4e2zaBODmcaUQhkd8vjDpX5tkw1f7i5qDwfBwpL05O7gwAOjD1vew

**Reply by harmonyball** [reply]
I agree that Friendly URL's have nothing to do with ranking and having read anumber of opinions including those by Torvista I now think that they likely should remain as plugins, however prominently marketed by ZenCart as readily available, easily installed and reliable etc. etc.
The Semrush article is very interesting - if the author is even close to the mark it confirms my experience that Page Speed is not as critical as Google would have us think - in one of my niche's a competitor sitting between 1 and 4 in rankings, dependent on search query, consistently fails the page speed performance tests with scores in the teens! (Lighthouse, Geometrix) - consitently over more than a year, but Google cares not and continues to rank them high. The green 'User Experience Factors' are few and far between and way down the pecking order (makes a mockery of the basis of my earlier comment) and to think that I was obsessed with page speed for a long time (achieving 97% plus on Mobile) but to no end as far as ranking is concerned.

**Comment by dbltoe** [comment]
And, looking at the above, I also found something many of my customers could use.  We already have Featured Products, but what about adding [Featured Categories](https://www.semrush.com/blog/seo-ecommerce-category-pages/)?

**Reply by scottcwilson** [reply]
To prevent thread drift, would you please create a new discussion on this topic?  Then I'll remove this reply.

**Comment by webconcept-one** [comment]
These kind of "friendly" URL become very unfriendly when the shop has some substancial amount of subcategories.

Instead of a short dynamic URL we get a horrible long "friendly" URL.

As dbltoe says there really are other functions which would improve the shop quality IMHO

**Comment by torvista** [comment]
Who cares if it improves SEO or not? answer: NEW/POTENTIAL USERS (who don't know any better, yet), but who are desperately needed.
I say add USU to core, (disabled), so it can be added to the feature list **so Zen Cart can compete again**.

**Comment by Motovista** [comment]
Someone is confused. 

https://www.BIG_GUY.com/Nature-Made-Calcium-Magnesium-Vitamin/dp/B000OD1F7U?pd_rd_w=0x1eX&content-id=amzn1.sym.1a072a44-84d1-4c51-bf0f-3547594e8217&pf_rd_p=1a072a44-84d1-4c51-bf0f-3547594e8217&pf_rd_r=DSQWZZE97BXP3PBZ3KDZ&pd_rd_wg=MoG2O&pd_rd_r=aca81c85-68a6-48e5-814f-a98d4fe81dc6&pd_rd_i=B000OD1F7U&psc=1&ref_=pd_bap_d_grid_rp_0_1_ec_i is not the URL.

https://www.BIG_GUY.com/Nature-Made-Calcium-Magnesium-Vitamin/dp/B000OD1F7U is the url. The rest of the numbers and letters are unique to this particular search. Instead of debating back and forth ad naseum, and quoting the first "expert" who sort of agrees with you, ask google what they want to see. 

https://developers.google.com/search/docs/crawling-indexing/url-structure

After 15 years, I'm looking to bail from Zencart because  they don't have a native SEO friendly URL structure.  When the simple seo url module quit working for us, it likely cost us thousands of dollars in lost sales. Now I'm going to try the CEON one. The main reason I haven't switched is that we have thousands of products and I haven't found another platform with the same sort of Multiple Category Link Manager function as ZenCart. If there is one, please let me know.

**Reply by lat9** [reply]
I'll note that Ultimate SEO URLs (USU) (https://www.zen-cart.com/downloads.php?do=file&id=132) is _similar_ in its setup/usage to the Simple SEO URLs.

**Reply by webconcept-one** [reply]
I'm sorry for my English. It isn't too easy to follow every of your details. I ll try.

The Google link you send is interesting.
I'd say zen-cart's url beauty factor is around a 5/10.

At an certain Shop with 2000 products and daily 400 users there was enough data. For me the data shows that Google Crawlers had been able ago to index deeper sub-categories (tested up to 5) already many years ago.

I can say that CEON URL installation is very simple, well documented and CEON URL really just works perfectly. I can't say anything about other URL plugins.
It had very little to zero impact on SEO.

I understand when anyone wants to use Friendly URL. I really do.

The question ist not whether a certain function or a plug-in are useful or not.
The question is whether it would be better or worse to implement it to the core.

Easier zen-cart upgrades to newest version. No extra plug-in needed. -> PRO
Higher maintanance time, higher complexity (large .htaccess files; additional rules), in general code that make the modular and most important base package bigger -> CON

**Reply by scottcwilson** [reply]
> When the simple seo url module quit working for us, it likely cost us thousands of dollars in lost sales. 

That's a huge cost.  What was your process for finding a contractor to fix it?   Why do you believe you were unsuccessful in getting this done?

**Comment by torvista** [comment]
Bump.
Why?
My simple search indicates that in 2026 friendly urls now do have a positive SEO effect (Google rewards more human-friendly content).
**Regardless of that**, more importantly new users expect it, so it should be in the core feature list. Period.

My approach for the path of least resistance would be to include USU in the core/as a core plugin, with a ZC docs section that details the url format it produces. This does not upset any apple carts/ticks the feature box/is a no brainer, in my view.

Additionally, in the docs, include current info about the CEON alternative if users want total control over URL format. I can write that.

(I'm getting jaded doing the maintenance **alone** on CEON URI + UMM version + UM Manager, BUT I don't like much the USU url format so that's my personal beef. Ceon Support take note and perhaps clarify the current status, to me at least.)

**Reply by sloanxue** [reply]
Ceon URI is definitely no worse than USU. It’s just my bad that I couldn’t find it easily. So just integrate it please! Users will never complain about having too many features in Zen Cart. LOL

**Reply by torvista** [reply]
USU is written and maintained by a core ZC developer.
It is not currently an encapsulated plugin so converting that is an additional non-trivial task which must be justified by the possible benefits of having this as a core feature. Who's paying for that dev time?
And, no licensing issues.

Meanwhile CEON Uri Mapping is a far more complex animal.
It exists 
a) as a free add-on, 
b) then there is a comercial UMM version which extends the auto-generation functionality/has modified versions of the same files used in a)  and adds an Admin management console: UMM Manager. 100% worth the money and I would encourage anyone to upgrade to that.

So, three sets of files, a licensing issue and haphazard support if at all from whoever is tasked with CEON support. I've kept it alive for years and have repos for all three. A thankless task.
Nonethless is it an awesome pile of work which offers total control of the urls and deserves to live on (RIP Conor the developer).

Converting that beast to a encapulsated plugin!..."show me the money", then who owns it (Conor's family), who's to support it...so, integrated into the core? Not going to happen.

So, in my view there's only one option and one decision to make about it.

**Comment by sloanxue** [comment]
Many thanks to the Zen Cart team and all plugin developers and maintainers! ZC 2.2.2 is absolutely amazing!
I’m not just here to express my gratitude, but also share my personal experience of using it.
Recently, I’ve been building a brand-new website based on Zen Cart 2.2.2. I have installed ZCA Bootstrap-4 Version 3.7.10, ga4-analytics-2.0.2, news_box_manager_v3-3.2.1, one_page_checkout-2.6.1, sitemapxml-4.0.5, usu-4.0.2, Zen_Cart-Structured_Data-2.1.0 and zen_Image-Handler-5.3.6. Both the templates and plugins work perfectly.
During the setup process, I even considered giving up ZC 2.2.2 and switching back to ZC 1.5.7. As a long-term user, I spent almost a whole day just searching for and installing various plugins.
The biggest trouble lies in finding suitable plugins. Just imagine how tough it would be for new users; most of them would easily get discouraged and quit. Quite a number of veteran users have already left for this reason.
I have no intention of arguing over which plugins should be integrated into the core system. I fully understand that every user has different functional demands, and not everyone needs all these plugins.
Nevertheless, no one would mind having more optional plugins, since users can simply disable the ones they don’t need. Hardly anyone will abandon Zen Cart just because it comes with a few unnecessary built-in features.
On the contrary, the lack of plugins, difficulty finding resources and complicated installation steps will scare off new users and drive away loyal old users.
Why are templates and plugins kept so hard to find? The search function for plugins on the official ZC forum is inefficient, and many resources are not updated timely. I finally had to look for plugins on GitHub instead.
Dear plugin developers and maintainers, you devote plenty of time and energy to developing these useful functions and keep updating them constantly. Why not gather all your excellent works in one place? Hiding such great resources is truly a waste.
As ordinary users, we care little about core system architecture. What we really need is comprehensive functions and easy operation.
Therefore, I sincerely suggest the Zen Cart team stick to your original core development philosophy, and meanwhile release an all-in-one version pre-installed with mainstream practical plugins and modern templates alongside each official core release.
I hesitated for a very long time before upgrading from ZC 1.5.7 to 2.2.2, mainly worried about insufficient templates and plugins. I even once planned to switch to WordPress or Shopify.
If the new official version comes with commonly-used plugins and mature Bootstrap templates, plenty of old users will definitely come back happily.
Once again, thank you sincerely to the entire Zen Cart team and all plugin developers, especially Cindy @lat9. You have almost every plugin I need, I really appreciate your great work!
